src/main/java/io/github/fablabsmc/fablabs/api/fluidvolume/v1/math/Fraction.java

```diff
@@ -209,55 +136,172 @@ public static Fraction min(Fraction left, Fraction right) {
+	}
+
+	public Fraction inverse() throws ArithmeticException {
+		// don't need to simplify
+		switch (this.signum()) {
+		case 1:
+			return new Fraction(this.denominator, this.numerator);
+		case -1:
+			return new Fraction(-this.denominator, -this.numerator);
+		default:
+			throw new ArithmeticException("Cannot invert zero fraction!");
+		}
+	}
+
+	public int signum() {
+		return Integer.signum(this.numerator);
+	}
+
+	public Fraction floorNearest(Fraction fraction) {
+		return fraction.multiply(ofWhole(this.floorDiv(fraction)));
+	}
+
+	public int floorDiv(Fraction fraction) {
+		return Math.floorDiv(this.numerator * fraction.denominator, fraction.numerator * this.denominator);
+	}
+
+	public boolean isNegative() {
+		return this.numerator < 0;
+	}
+
+	public boolean isPositive() {
+		return this.numerator > 0;
+	}
+
 	@Override
-	public int intValue() {
-		return numerator / denominator;
+	public double doubleValue() {
+		return ((double) this.numerator) / this.denominator;
+	}
+
+	public Fraction negate() {
+		// don't need to simplify
+		return new Fraction(-this.numerator, this.denominator);
+	}
+
+	public Fraction multiply(int amount) {
+		return of(this.numerator * amount, this.denominator);
+	}
+
+	public Fraction add(Fraction other) {
+		int commonMultiple = lcm(this.denominator, other.denominator);
+		int leftNumerator = commonMultiple / this.denominator * this.numerator;
+		int rightNumerator = commonMultiple / other.denominator * other.numerator;
+		return ofValidDenominator(leftNumerator + rightNumerator, commonMultiple);
+	}
+
+	public Fraction subtract(Fraction other) {
+		int commonMultiple = lcm(this.denominator, other.denominator);
+		int leftNumerator = commonMultiple / this.denominator * this.numerator;
+		int rightNumerator = commonMultiple / other.denominator * other.numerator;
+		return ofValidDenominator(leftNumerator - rightNumerator, commonMultiple);
+	}
+
+	public Fraction floorWithDenominator(int denom) {
+		Preconditions.checkArgument(denom > 0, "New denominator must be positive!");
+		if (denom == this.getDenominator()) return this;
+		return Fraction.ofValidDenominator(Math.floorDiv(this.numerator * denom, this.denominator), denom);
+	}
+
+	public boolean isGreaterThan(Fraction fraction) {
+		return this.compareTo(fraction) > 0;
 	}
 
 	@Override
-	public long longValue() {
-		return intValue();
+	public int compareTo(Fraction other) {
+		long leftNum = this.numerator * other.denominator;
+		long rightNum = other.numerator * this.denominator;
+		return Long.compare(leftNum, rightNum);
+	}
+
+	public boolean isGreaterThanOrEqualTo(Fraction fraction) {
+		return this.compareTo(fraction) >= 0;
 	}
 
 	@Override
-	public float floatValue() {
-		return (float) doubleValue();
+	public int hashCode() {
+		int result = this.numerator;
+		result = 31 * result + this.denominator;
```

Comment:
Should be a lot harder to collide now, and we can probably change it in the future if necessary since nobody cares about the hard value.

Reply:
the value is intellij 😛, which comes from "effective java" or something

Reply:
Ahh, if idea is suggesting this then I guess we can trust it for now.